// threepipe/lib/core/light/RectAreaLight2.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { RectAreaLight } from 'three';
import { iLightCommons } from '../object/iLightCommons';
import { uiColor, uiNumber, uiPanelContainer, uiSlider, uiToggle } from 'uiconfig.js';
import { onChange3 } from 'ts-browser-helpers';
let RectAreaLight2 = class RectAreaLight2 extends RectAreaLight {
    constructor(color, intensity, width, height) {
        super(color, intensity, width, height);
        this.assetType = 'light';
        this.setDirty = iLightCommons.setDirty;
        this.refreshUi = iLightCommons.refreshUi;
        this.isRectAreaLight2 = true;
        iLightCommons.upgradeLight.call(this);
    }
    autoScale() {
        console.warn('AutoScale not supported on Lights');
        return this;
    }
    autoCenter() {
        console.warn('AutoCenter not supported on Lights');
        return this;
    }
    /**
     * @deprecated use `this` instead
     */
    get lightObject() {
        return this;
    }
    /**
     * @deprecated use `this` instead
     */
    get modelObject() {
        return this;
    }
};
__decorate([
    uiToggle('Enabled'),
    onChange3('setDirty')
], RectAreaLight2.prototype, "visible", void 0);
__decorate([
    uiColor('Color', (that) => ({ onChange: () => that.setDirty() }))
], RectAreaLight2.prototype, "color", void 0);
__decorate([
    uiSlider('Intensity', [0, 30], 0.01),
    onChange3('setDirty')
], RectAreaLight2.prototype, "intensity", void 0);
__decorate([
    uiNumber('Width'),
    onChange3('setDirty')
], RectAreaLight2.prototype, "width", void 0);
__decorate([
    uiNumber('Height'),
    onChange3('setDirty')
], RectAreaLight2.prototype, "height", void 0);
__decorate([
    uiNumber('Power'),
    onChange3('setDirty')
], RectAreaLight2.prototype, "power", void 0);
RectAreaLight2 = __decorate([
    uiPanelContainer('RectArea Light')
], RectAreaLight2);
export { RectAreaLight2 };
